File: ios/RunFormCoachAI/GaitAnalyzer.swift
import Foundation
import os.log

// MARK: - GaitAnalyzer

/// Real-time gait metric extractor using fused accelerometer + gyroscope data.
///
/// Computes from a sliding window of `SensorFrame` samples:
/// - **Vertical oscillation** (cm): Double-integrate vertical acceleration after gravity removal.
/// - **Ground contact time** (ms): Estimated from gyroscope pitch-rate zero-crossing during stance.
/// - **Trunk lean angle** (degrees): Tilt of the gravity vector in the sagittal plane.
///
/// Uses `SignalProcessing` pure functions for smoothing and statistics.
/// Biomechanical formulas are aligned with `PoseExtractor`'s video-based analysis.
public final class GaitAnalyzer: @unchecked Sendable {

    // MARK: - Configuration

    /// Analysis window duration in seconds.
    public let windowSeconds: TimeInterval

    /// Assumed sampling rate (Hz) — used for integration time-step.
    public let samplingRate: Double

    /// Gravity magnitude threshold for lean estimation (g).
    private let gravityThreshold: Double = 0.85

    // MARK: - State

    /// Most recent gait snapshot.
    public private(set) var currentSnapshot: GaitSnapshot?

    /// Latest cadence SPM value, injected by RunSessionManager from CadenceDetector.
    public var latestCadenceSPM: Double = 0.0

    /// Rolling statistics for the current window.
    public private(set) var verticalOscillationStats: (mean: Double, stdDev: Double, trend: Double)?

    /// Callback invoked when a new gait snapshot is computed.
    public var onGaitUpdate: (@Sendable (GaitSnapshot) -> Void)?

    // MARK: - Private

    private let queue = DispatchQueue(label: "com.runformcoachai.gait")
    private var accelZHistory: [Double] = []       // vertical acceleration
    private var accelYHistory: [Double] = []       // forward acceleration (for lean)
    private var gyroXHistory: [Double] = []        // pitch rate (for GCT)
    private var timestamps: [TimeInterval] = []
    private var frameCount: Int = 0
    private var lastSnapshotTime: Date?

    // Integration state for vertical oscillation
    private var velocityZ: Double = 0.0
    private var positionZ: Double = 0.0
    private var positionZHistory: [Double] = []
    private var lastTimestamp: TimeInterval?

    // Gravity estimate (slow-moving baseline of accelZ)
    private var gravityBaseline: Double = -1.0

    private var maxHistorySamples: Int {
        max(Int(windowSeconds * samplingRate), 60)
    }

    // MARK: - Init

    /// - Parameters:
    ///   - windowSeconds: Analysis window (default 5.0, range 3–10).
    ///   - samplingRate: Expected sensor rate in Hz (default 60).
    public init(windowSeconds: TimeInterval = 5.0, samplingRate: Double = 60) {
        self.windowSeconds = max(3, min(10, windowSeconds))
        self.samplingRate = samplingRate
    }

    // MARK: - Public API

    /// Feed a single `SensorFrame` and optionally produce a gait update.
    /// - Parameter frame: The sensor data point.
    public func process(frame: SensorFrame) {
        queue.async { [weak self] in
            self?.processSync(frame)
        }
    }

    /// Process a batch of frames (e.g., from ring buffer snapshot).
    /// - Parameter frames: Array of sensor frames in time order.
    public func processBatch(_ frames: [SensorFrame]) {
        queue.async { [weak self] in
            for frame in frames {
                self?.processSync(frame)
            }
        }
    }

    /// Reset all accumulated state.
    public func reset() {
        queue.sync {
            accelZHistory.removeAll(keepingCapacity: true)
            accelYHistory.removeAll(keepingCapacity: true)
            gyroXHistory.removeAll(keepingCapacity: true)
            timestamps.removeAll(keepingCapacity: true)
            positionZHistory.removeAll(keepingCapacity: true)
            frameCount = 0
            velocityZ = 0.0
            positionZ = 0.0
            lastTimestamp = nil
            gravityBaseline = -1.0
            currentSnapshot = nil
            verticalOscillationStats = nil
            lastSnapshotTime = nil
        }
    }

    // MARK: - Private processing

    private func processSync(_ frame: SensorFrame) {
        frameCount += 1

        // Append to history buffers
        accelZHistory.append(frame.accelerationZ)
        accelYHistory.append(frame.accelerationY)
        gyroXHistory.append(frame.rotationRateX)
        timestamps.append(frame.timestamp)

        // Enforce window
        while accelZHistory.count > maxHistorySamples {
            accelZHistory.removeFirst()
            accelYHistory.removeFirst()
            gyroXHistory.removeFirst()
            timestamps.removeFirst()
            if !positionZHistory.isEmpty { positionZHistory.removeFirst() }
        }

        // --- Vertical Oscillation via double-integration ---

        // Update gravity baseline with slow exponential smoothing
        let alphaGrav = 0.95  // slow adaptation
        gravityBaseline = alphaGrav * gravityBaseline + (1.0 - alphaGrav) * frame.accelerationZ

        // Remove gravity → dynamic vertical acceleration
        let dynamicAccelZ = frame.accelerationZ - gravityBaseline

        // Integrate: v[n] = v[n-1] + a[n] * dt
        if let lastTS = lastTimestamp {
            let dt = max(0.001, frame.timestamp - lastTS)
            velocityZ += dynamicAccelZ * dt

            // High-pass filter velocity to remove drift (τ = 1.5s)
            let alphaVel = exp(-dt / 1.5)
            velocityZ *= alphaVel

            // Integrate velocity → position
            positionZ += velocityZ * dt

            // Drift compensation: leak position toward zero slowly
            positionZ *= 0.998
        }
        lastTimestamp = frame.timestamp

        positionZHistory.append(positionZ)
        while positionZHistory.count > maxHistorySamples {
            positionZHistory.removeFirst()
        }

        // --- Trunk Lean Angle ---

        // From accelerometer: lean ≈ atan2(accelY, -accelZ) when phone is portrait
        // accelY = forward, accelZ = up (negative when upright due to gravity)
        // lean > 0 = forward lean
        let smoothedAccelY = exponentialSmooth(accelYHistory, alpha: 0.9)
        let smoothedAccelZ = exponentialSmooth(accelZHistory, alpha: 0.9)
        let leanRadians = atan2(smoothedAccelY, -smoothedAccelZ)
        let trunkLeanDeg = leanRadians * 180.0 / .pi

        // --- Ground Contact Time ---

        // Estimated from gyro pitch-rate pattern:
        // During foot strike, there's a rapid pitch deceleration spike.
        // GCT ≈ width of gyro spike above threshold.
        let gctEstimate = estimateGCT(from: gyroXHistory)

        // --- Compute snapshot ---
        let vertOscCm = computeVerticalOscillation()

        // Update rolling statistics
        if positionZHistory.count >= 10 {
            let meanPos = positionZHistory.reduce(0, +) / Double(positionZHistory.count)
            let stdPos = stdDev(positionZHistory)
            // Trend: slope of linear regression on last N samples
            let trend = computeTrend(positionZHistory)
            verticalOscillationStats = (mean: meanPos, stdDev: stdPos, trend: trend)
        }

        // Gentle per-frame velocity decay prevents unbounded drift without visible jumps.
        // 0.9995 per frame at 60 Hz ≈ 3% decay per second — smooth and imperceptible.
        velocityZ *= 0.9995

        // Emit snapshot (throttled: at most every 1.0s)
        let now = Date()
        if let last = lastSnapshotTime, now.timeIntervalSince(last) < 1.0 {
            return
        }
        lastSnapshotTime = now

        let cadenceForSnapshot: Double = latestCadenceSPM

        let snapshot = GaitSnapshot(
            verticalOscillationCm: vertOscCm,
            groundContactTimeMs: gctEstimate,
            trunkLeanDegrees: trunkLeanDeg,
            cadenceSPM: cadenceForSnapshot,
            timestamp: now
        )
        currentSnapshot = snapshot
        onGaitUpdate?(snapshot)
    }

    // MARK: - Metric computations

    /// Vertical oscillation amplitude from position history (peak-to-peak / 2, in cm).
    /// G-force to cm conversion: 1g ≈ 9.81 m/s², acceleration double-integrated.
    private func computeVerticalOscillation() -> Double {
        guard positionZHistory.count >= 10 else { return 0 }

        // positionZ is in meters (g * s² → multiply by 9.81 to get meters)
        let positionsM = positionZHistory.map { $0 * 9.81 }
        let pMin = positionsM.min() ?? 0
        let pMax = positionsM.max() ?? 0
        let amplitudeM = (pMax - pMin) / 2.0
        let amplitudeCm = amplitudeM * 100.0

        return SignalProcessing.clamp(amplitudeCm, 0, 35)
    }

    /// Estimate ground contact time from gyroscope pitch-rate signal.
    ///
    /// During stance, the foot acts as a pivot producing a rapid pitch rotation.
    /// GCT ≈ duration the gyro magnitude stays above 30% of its peak.
    private func estimateGCT(from gyroHistory: [Double]) -> Double {
        guard gyroHistory.count >= 10 else { return 0 }

        let absGyro = gyroHistory.map { abs($0) }
        guard let peak = absGyro.max(), peak > 0.1 else { return 0 }

        let threshold = peak * 0.3
        var aboveThreshold = 0
        for val in absGyro where val >= threshold {
            aboveThreshold += 1
        }

        // Convert sample count to milliseconds
        let durationSec = Double(aboveThreshold) / samplingRate
        let durationMs = durationSec * 1000.0

        // Typical GCT for runners: 150–300ms
        return SignalProcessing.clamp(durationMs, 80, 500)
    }

    /// Exponential moving average of the last value in history.
    private func exponentialSmooth(_ history: [Double], alpha: Double) -> Double {
        guard !history.isEmpty else { return 0 }
        guard history.count > 1 else { return history[0] }
        // Use the latest value as anchor, blend slightly with history
        let latest = history.last!
        let prevAvg = history.dropLast().reduce(0, +) / Double(history.count - 1)
        return alpha * prevAvg + (1.0 - alpha) * latest
    }

    /// Simple trend: sign of difference between recent and older halves.
    private func computeTrend(_ values: [Double]) -> Double {
        let n = values.count
        guard n >= 8 else { return 0 }
        let mid = n / 2
        let olderMean = values[0..<mid].reduce(0, +) / Double(mid)
        let recentMean = values[mid..<n].reduce(0, +) / Double(n - mid)
        return recentMean - olderMean
    }

    /// Population standard deviation.
    private func stdDev(_ values: [Double]) -> Double {
        let n = Double(values.count)
        guard n > 1 else { return 0 }
        let m = values.reduce(0, +) / n
        let variance = values.reduce(0) { $0 + ($1 - m) * ($1 - m) } / n
        return sqrt(variance)
    }
}

File: android/app/src/main/java/com/runformcoach/runformcoachai/sensor/GaitAnalyzer.kt
package com.runformcoach.runformcoachai.sensor

import android.util.Log
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.exp
import kotlin.math.sqrt

// ── GaitSnapshot ─────────────────────────────────────────────────────────────

/**
 * 实时步态快照，由 [GaitAnalyzer] 产出。
 *
 * @property verticalOscillationCm  垂直振幅 (cm)
 * @property groundContactTimeMs    触地时间估算 (ms)
 * @property trunkLeanDegrees       躯干倾角 (°), 正值 = 前倾
 * @property cadenceSPM             快照时的滚动平均步频
 * @property timestamp              快照时间戳 (epoch millis)
 */
data class GaitSnapshot(
    val verticalOscillationCm: Double = 0.0,
    val groundContactTimeMs: Double = 0.0,
    val trunkLeanDegrees: Double = 0.0,
    val cadenceSPM: Double = 0.0,
    val timestamp: Long = System.currentTimeMillis()
)

/**
 * 垂直振幅滚动统计。
 */
data class VerticalOscillationStats(
    val mean: Double,
    val stdDev: Double,
    val trend: Double
)

// ── GaitAnalyzer ─────────────────────────────────────────────────────────────

/**
 * 实时步态指标提取器 —— 加速度计 + 陀螺仪融合。
 *
 * 从 [SensorFrame] 滑动窗口计算:
 * - **垂直振幅** (cm): 垂直加速度去除重力后双重积分
 * - **触地时间** (ms): 从陀螺仪 pitch-rate 零交叉估算
 * - **躯干倾角** (°): 矢状面重力向量倾角
 *
 * 生物力学公式对齐视频分析 [PoseExtractor]。
 *
 * 输出 [StateFlow]<[GaitSnapshot]?> 供 UI 订阅。
 *
 * 用法:
 * ```
 * val analyzer = GaitAnalyzer(windowSeconds = 5.0, samplingRate = 60.0)
 * analyzer.process(sensorFrame)
 * analyzer.currentSnapshot.collect { snapshot -> ... }
 * ```
 */
class GaitAnalyzer(
    /** 分析窗口时长 (秒)，限制 3–10。 */
    windowSeconds: Double = 5.0,
    /** 假定采样率 (Hz) —— 用于积分时间步长。 */
    val samplingRate: Double = 60.0
) {
    companion object {
        private const val TAG = "GaitAnalyzer"
        private const val GRAVITY_THRESHOLD = 0.85
        /** 每次发射快照的最小间隔 (秒)。 */
        private const val EMIT_THROTTLE_SEC = 1.0
        /** 速度积分器衰减周期 (约每 30 秒样本数)。 */
        private const val VELOCITY_DECAY_PERIOD_SAMPLES = 1800  // 30s * 60 Hz
    }

    /** 分析窗口时长 (秒)。 */
    val windowSeconds: Double = maxOf(3.0, minOf(10.0, windowSeconds))

    // ── 公开状态 ──────────────────────────────────────────────────────────────
    private val _currentSnapshot = MutableStateFlow<GaitSnapshot?>(null)

    /// Latest cadence SPM, injected by RunSessionManager from CadenceDetector.
    @Volatile var latestCadenceSPM: Double = 0.0
    val currentSnapshot: StateFlow<GaitSnapshot?> = _currentSnapshot.asStateFlow()

    private val _verticalOscillationStats = MutableStateFlow<VerticalOscillationStats?>(null)
    val verticalOscillationStats: StateFlow<VerticalOscillationStats?> = _verticalOscillationStats.asStateFlow()

    // ── 内部状态 ──────────────────────────────────────────────────────────────

    private val accelZHistory = ArrayDeque<Double>()
    private val accelYHistory = ArrayDeque<Double>()
    private val gyroXHistory = ArrayDeque<Double>()
    private val timestamps = ArrayDeque<Long>()
    private var frameCount: Int = 0
    private var lastSnapshotEmitTime: Long = 0L

    // 垂直振幅积分状态
    private var velocityZ: Double = 0.0
    private var positionZ: Double = 0.0
    private val positionZHistory = ArrayDeque<Double>()
    private var lastTimestampNanos: Long? = null

    // 重力基线 (加速度 Z 轴的慢速指数移动平均)
    private var gravityBaseline: Double = -1.0

    private val maxHistorySamples: Int
        get() = maxOf((windowSeconds * samplingRate).toInt(), 60)

    // ── 公开 API ──────────────────────────────────────────────────────────────

    /** 喂入单个 [SensorFrame]，可能产出一个步态更新。 */
    fun process(frame: SensorFrame) {
        processSync(frame)
    }

    /** 批量处理 (例如环形缓冲区快照)。 */
    fun processBatch(frames: List<SensorFrame>) {
        for (frame in frames) {
            processSync(frame)
        }
    }

    /** 重置所有累积状态。 */
    fun reset() {
        accelZHistory.clear()
        accelYHistory.clear()
        gyroXHistory.clear()
        timestamps.clear()
        positionZHistory.clear()
        frameCount = 0
        velocityZ = 0.0
        positionZ = 0.0
        lastTimestampNanos = null
        gravityBaseline = -1.0
        _currentSnapshot.value = null
        _verticalOscillationStats.value = null
        lastSnapshotEmitTime = 0L
    }

    // ── 内部处理 ──────────────────────────────────────────────────────────────

    private fun processSync(frame: SensorFrame) {
        frameCount++

        // 追加到历史缓冲区 (使用 accelZ 作为垂直, accelY 作为前后, gyroX 为 pitch rate)
        accelZHistory.addLast(frame.accelZ.toDouble())
        accelYHistory.addLast(frame.accelY.toDouble())
        gyroXHistory.addLast(frame.gyroX.toDouble())
        timestamps.addLast(frame.timestampNanos)

        // 强制窗口大小
        while (accelZHistory.size > maxHistorySamples) {
            accelZHistory.removeFirst()
            accelYHistory.removeFirst()
            gyroXHistory.removeFirst()
            timestamps.removeFirst()
            if (positionZHistory.isNotEmpty()) positionZHistory.removeFirst()
        }

        // ── 垂直振幅 (双重积分) ─────────────────────────────────────────────

        // 更新重力基线 (慢速 EMA)
        val alphaGrav = 0.95
        gravityBaseline = alphaGrav * gravityBaseline + (1.0 - alphaGrav) * frame.accelZ

        // 去除重力 → 动态垂直加速度
        val dynamicAccelZ = frame.accelZ - gravityBaseline

        // 积分: v[n] = v[n-1] + a[n] * dt
        if (lastTimestampNanos != null) {
            val dtSec = maxOf(0.001, (frame.timestampNanos - lastTimestampNanos!!) / 1_000_000_000.0)
            velocityZ += dynamicAccelZ * dtSec

            // 高通滤波速度去除漂移 (τ = 1.5s)
            val alphaVel = exp(-dtSec / 1.5)
            velocityZ *= alphaVel

            // 积分速度 → 位移
            positionZ += velocityZ * dtSec

            // 漂移补偿: 缓慢泄露位移趋向 0
            positionZ *= 0.998
        }
        lastTimestampNanos = frame.timestampNanos

        positionZHistory.addLast(positionZ)
        while (positionZHistory.size > maxHistorySamples) {
            positionZHistory.removeFirst()
        }

        // ── 躯干倾角 ────────────────────────────────────────────────────────

        // 从加速度计估算: lean ≈ atan2(accelY, -accelZ)
        // accelY = 前后, accelZ = 垂直 (正值向上时, 直立时 accelZ ≈ -1g)
        val smoothedAccelY = exponentialSmoothLast(accelYHistory, alpha = 0.9)
        val smoothedAccelZ = exponentialSmoothLast(accelZHistory, alpha = 0.9)
        val leanRadians = atan2(smoothedAccelY, -smoothedAccelZ)
        val trunkLeanDeg = leanRadians * 180.0 / Math.PI

        // ── 触地时间 (GCT) ──────────────────────────────────────────────────

        val gctEstimate = estimateGCT(gyroXHistory)

        // ── 垂直振幅计算 ────────────────────────────────────────────────────

        val vertOscCm = computeVerticalOscillation()

        // ── 滚动统计 ────────────────────────────────────────────────────────

        if (positionZHistory.size >= 10) {
            val meanPos = positionZHistory.sum() / positionZHistory.size
            val stdPos = stdDev(positionZHistory)
            val trend = computeTrend(positionZHistory)
            _verticalOscillationStats.value = VerticalOscillationStats(
                mean = meanPos, stdDev = stdPos, trend = trend
            )
        }

        // 定期衰减速度积分器，防止长时间运行中漂移无限增长
        if (frameCount % VELOCITY_DECAY_PERIOD_SAMPLES == 0) {
            velocityZ *= 0.5
        }

        // ── 限流发射 (最多 1 Hz) ───────────────────────────────────────────

        val now = System.currentTimeMillis()
        val throttleMs = (EMIT_THROTTLE_SEC * 1000).toLong()
        if (now - lastSnapshotEmitTime < throttleMs) {
            return
        }
        lastSnapshotEmitTime = now

        val snapshot = GaitSnapshot(
            verticalOscillationCm = vertOscCm,
            groundContactTimeMs = gctEstimate,
            trunkLeanDegrees = trunkLeanDeg,
            cadenceSPM = latestCadenceSPM,
            timestamp = now
        )
        _currentSnapshot.value = snapshot
    }

    // ── 指标计算 ──────────────────────────────────────────────────────────────

    /**
     * 从位移历史计算垂直振幅 (峰峰值 / 2, 单位 cm)。
     *
     * positionZ 单位: g·s² → ×9.81 m/s² → m → ×100 cm。
     */
    private fun computeVerticalOscillation(): Double {
        if (positionZHistory.size < 10) return 0.0
        val positionsM = positionZHistory.map { it * 9.81 }
        val pMin = positionsM.minOrNull() ?: 0.0
        val pMax = positionsM.maxOrNull() ?: 0.0
        val amplitudeM = (pMax - pMin) / 2.0
        val amplitudeCm = amplitudeM * 100.0
        return clamp(amplitudeCm, 0.0, 35.0)
    }

    /**
     * 从陀螺仪 pitch-rate 信号估算触地时间。
     *
     * 支撑期脚部作为支点产生快速俯仰旋转。
     * GCT ≈ 陀螺仪幅值超过峰值 30% 的样本数 / 采样率。
     */
    private fun estimateGCT(gyroHistory: List<Double>): Double {
        if (gyroHistory.size < 10) return 0.0
        val absGyro = gyroHistory.map { abs(it) }
        val peak = absGyro.maxOrNull() ?: 0.0
        if (peak <= 0.1) return 0.0

        val threshold = peak * 0.3
        var aboveThreshold = 0
        for (v in absGyro) {
            if (v >= threshold) aboveThreshold++
        }
        val durationSec = aboveThreshold / samplingRate
        val durationMs = durationSec * 1000.0
        // 典型跑步者 GCT: 150–300 ms
        return clamp(durationMs, 80.0, 500.0)
    }

    /** 对历史序列最后值的指数移动平均。 */
    private fun exponentialSmoothLast(history: List<Double>, alpha: Double): Double {
        if (history.isEmpty()) return 0.0
        if (history.size == 1) return history[0]
        val latest = history.last()
        val prevAvg = history.dropLast(1).sum() / (history.size - 1)
        return alpha * prevAvg + (1.0 - alpha) * latest
    }

    /** 简单趋势: 近期一半与较旧一半均值之差。 */
    private fun computeTrend(values: List<Double>): Double {
        val n = values.size
        if (n < 8) return 0.0
        val mid = n / 2
        val olderMean = values.subList(0, mid).sum() / mid
        val recentMean = values.subList(mid, n).sum() / (n - mid)
        return recentMean - olderMean
    }

    /** 总体标准差。 */
    private fun stdDev(values: List<Double>): Double {
        val n = values.size.toDouble()
        if (n <= 1) return 0.0
        val m = values.sum() / n
        val variance = values.sumOf { (it - m) * (it - m) } / n
        return sqrt(variance)
    }

    private fun clamp(value: Double, min: Double, max: Double): Double {
        return value.coerceIn(min, max)
    }
}
